refactor(dev): task-queue based deisgn

hyf0 opened the pull request:
In the previous design, it’s hard to think about how to handle complex cases. In this design, by abstract each change into a queued task, it makes parallel events sequential, which reduces the complexity of thinking itself.

hyf0 commented:
* **#6233** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6233?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6233?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6232** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6232?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6233).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hyf0 commented on crates/rolldown/src/dev/building_task.rs:
```diff
@@ -0,0 +1,161 @@
+    if let Err(err) = self.run_inner().await {
+      // FIXME: Should handle the error properly.
+      tracing::error!("Failed to run bundling task: {err:?}");
+      self.dev_context.state.lock().await.try_to_idle().expect("FIXME: Should not unwrap here");
```
I'm thinking how to handle cache if there's an error. But not this way.

hyf0 commented on crates/rolldown/src/dev/build_driver.rs:
```diff
@@ -31,56 +31,55 @@ impl BuildDriver {
-      // Note: Full rebuild and incremental build both clear changed files.
-      let changed_files = mem::take(&mut build_state.changed_files);
+
+      // TODO: hyf0 merge mergeable task inputs into one
```
Do we want the merge behavior? Or at least make it optional? @sapphi-red 

- Not merging ensures users could see newest changes if possible. Merging might cause a task that's able to do a successful hmr update fail.
- If we want to merge, the strategy would be a little sophisticated. If the current task is a [full] rebuild only task, it will merge all subsquent tasks until finding one that require to generate hmr update.

sapphi-red replied:
I think we want it. There won't be multiple hmr tasks queued unless the hmr is slow. This happens when there's a slow plugin. In that case, I think we should merge the tasks to avoid causing the hmr to be lagged behind. I think that has a benefit even with the downsides you described.
But this is an edge case because normally the HMR is instant and the watcher debounce will merge the file change events (i.e. this only happens if the HMR takes longer than the debounce tick rate). So I think we can leave it for now as long as the multiple rebuild tasks are merged (I assume this is done indirectly by the delay state). I think improving the error handling has a higher priority as that would affect more users.

hyf0 replied:
Got it. I'll start to think a solution for error handling.

hyf0 committed:
refactor(dev): task-queue based deisgn (#6233)

In the previous design, it’s hard to think about how to handle complex cases. In this design, by abstract each change into a queued task, it makes parallel events sequential, which reduces the complexity of thinking itself.
